# **Laboratorio 8: Ready, Set, Deploy! 👩‍🚀👨‍🚀**

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos - Otoño 2026 </strong></center>

### Cuerpo Docente:

- Profesores: Pablo Badilla, Diego Cortez
- Auxiliares: Melanie Peña, Valentina Rojas
- Ayudantes: Javiera Arévalo, Tamara Carrasco y Ignacio Reyes

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Javiera Yañez Sanchez

### **Link de repositorio de GitHub:** https://github.com/javiyansan/MDS7202

## Temas a tratar

- Entrenamiento y registro de modelos usando MLFlow.
- Despliegue de modelo usando FastAPI
- Containerización del proyecto usando Docker

### Objetivos principales del laboratorio

- Generar una solución a un problema a partir de ML
- Desplegar su solución usando MLFlow, FastAPI y Docker

El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

# **Introducción**

<p align="center">
  <img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExODJnMHJzNzlkNmQweXoyY3ltbnZ2ZDlxY2c0aW5jcHNzeDNtOXBsdCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/AbPdhwsMgjMjax5reo/giphy.gif" width="400">
</p>



Consumida en la tristeza el despido de Renacín, Smapina ha decaído en su desempeño, lo que se ha traducido en un irregular tratamiento del agua. Esto ha implicado una baja en la calidad del agua, llegando a haber algunos puntos de la comuna en la que el vital elemento no es apto para el consumo humano. Es por esto que la sanitaria pública de la municipalidad de Maipú se ha contactado con ustedes para que le entreguen una urgente solución a este problema (a la vez que dejan a Smapina, al igual que Renacín, sin trabajo 😔).

El problema que la empresa le ha solicitado resolver es el de elaborar un sistema que les permita saber si el agua es potable o no. Para esto, la sanitaria les ha proveido una base de datos con la lectura de múltiples sensores IOT colocados en diversas cañerías, conductos y estanques. Estos sensores señalan nueve tipos de mediciones químicas y más una etiqueta elaborada en laboratorio que indica si el agua es potable o no el agua.

La idea final es que puedan, en el caso que el agua no sea potable, dar un aviso inmediato para corregir el problema. Tenga en cuenta que parte del equipo docente vive en Maipú y su intoxicación podría implicar graves problemas para el cierre del curso.

Atributos:

1. pH value
2. Hardness
3. Solids (Total dissolved solids - TDS)
4. Chloramines
5. Sulfate
6. Conductivity
7. Organic_carbon
8. Trihalomethanes
9. Turbidity

Variable a predecir:

10. Potability (1 si es potable, 0 no potable)

Descripción de cada atributo se pueden encontrar en el siguiente link: [dataset](https://www.kaggle.com/adityakadiwal/water-potability)

# **1. Optimización de modelos con Optuna + MLFlow (2.0 puntos)**

El objetivo de esta sección es que ustedes puedan combinar Optuna con MLFlow para poder realizar la optimización de los hiperparámetros de sus modelos.

Como aún no hemos hablado nada sobre `MLFlow` cabe preguntarse: **¡¿Qué !"#@ es `MLflow`?!**

<p align="center">
  <img src="https://media.tenor.com/eusgDKT4smQAAAAC/matthew-perry-chandler-bing.gif" width="400">
</p>

## **MLFlow**

`MLflow` es una plataforma de código abierto que simplifica la gestión y seguimiento de proyectos de aprendizaje automático. Con sus herramientas, los desarrolladores pueden organizar, rastrear y comparar experimentos, además de registrar modelos y controlar versiones.

<p align="center">
  <img src="https://spark.apache.org/images/mlflow-logo.png" width="350">
</p>

Si bien esta plataforma cuenta con un gran número de herramientas y funcionalidades, en este laboratorio trabajaremos con dos:
1. **Runs**: Registro que constituye la información guardada tras la ejecución de un entrenamiento. Cada `run` tiene su propio run_id, el cual sirve como identificador para el entrenamiento en sí mismo. Dentro de cada `run` podremos acceder a información como los hiperparámetros utilizados, las métricas obtenidas, las librerías requeridas y hasta nos permite descargar el modelo entrenado.
2. **Experiments**: Se utilizan para agrupar y organizar diferentes ejecuciones de modelos (`runs`). En ese sentido, un experimento puede agrupar 1 o más `runs`. De esta manera, es posible también registrar métricas, parámetros y archivos (artefactos) asociados a cada experimento.

### **Todo bien pero entonces, ¿cómo se usa en la práctica `MLflow`?**

Es sencillo! Considerando un problema de machine learning genérico, podemos registrar la información relevante del entrenamiento ejecutando `mlflow.autolog()` antes entrenar nuestro modelo. Veamos este bonito ejemplo facilitado por los mismos creadores de `MLflow`:

```python
#!pip install mlflow
import mlflow # importar mlflow

from sklearn.model_selection import train_test_split
from sklearn.datasets import load_diabetes
from sklearn.ensemble import RandomForestRegressor

db = load_diabetes()
X_train, X_test, y_train, y_test = train_test_split(db.data, db.target)

# Create and train models.
rf = RandomForestRegressor(n_estimators=100, max_depth=6, max_features=3)

mlflow.autolog() # registrar automáticamente información del entrenamiento
with mlflow.start_run(): # delimita inicio y fin del run
    # aquí comienza el run
    rf.fit(X_train, y_train) # train the model
    predictions = rf.predict(X_test) # Use the model to make predictions on the test dataset.
    # aquí termina el run
```

Si ustedes ejecutan el código anterior en sus máquinas locales (desde un jupyter notebook por ejemplo) se darán cuenta que en su directorio *root* se ha creado la carpeta `mlruns`. Esta carpeta lleva el tracking de todos los entrenamientos ejecutados desde el directorio root (importante: si se cambian de directorio y vuelven a ejecutar el código anterior, se creará otra carpeta y no tendrán acceso al entrenamiento anterior). Para visualizar estos entrenamientos, `MLflow` nos facilita hermosa interfaz visual a la que podemos acceder ejecutando:

```
mlflow ui
```

y luego pinchando en la ruta http://127.0.0.1:5000 que nos retorna la terminal. Veamos en vivo algunas de sus funcionalidades!

<p align="center">
  <img src="https://media4.giphy.com/media/v1.Y2lkPTc5MGI3NjExZXVuM3A5MW1heDFpa21qbGlwN2pyc2VoNnZsMmRzODZxdnluemo2bCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/3o84sq21TxDH6PyYms/giphy.gif" width="400">
</p>

Les dejamos también algunos comandos útiles:

- `mlflow.create_experiment("nombre_experimento")`: Les permite crear un nuevo experimento para agrupar entrenamientos
- `mlflow.log_metric("nombre_métrica", métrica)`: Les permite registrar una métrica *custom* bajo el nombre de "nombre_métrica"


In [18]:
!uv add mlflow

  × No solution found when resolving dependencies for split (markers:               
  │ python_full_version >= '3.14' and sys_platform == 'win32'):
  ╰─▶ Because only the following versions of mlflow are available:
          mlflow<=2.13.0
          mlflow==2.13.1
          mlflow==2.13.2
          mlflow==2.14.0
          mlflow==2.14.1
          mlflow==2.14.2
          mlflow==2.14.3
          mlflow==2.15.0
          mlflow==2.15.1
          mlflow==2.16.0
          mlflow==2.16.1
          mlflow==2.16.2
          mlflow==2.17.0
          mlflow==2.17.1
          mlflow==2.17.2
          mlflow==2.18.0
          mlflow==2.19.0
          mlflow==2.20.0
          mlflow==2.20.1
          mlflow==2.20.2
          mlflow==2.20.3
          mlflow==2.20.4
          mlflow==2.21.0
          mlflow==2.21.1
          mlflow==2.21.2
          mlflow==2.21.3
          mlflow==2.22.0
          mlflow==2.22.1
          mlflow==2.22.2
          mlflow==2.22.3
          mlflow==2.22.4
         

Si tiene problemas puede necesitar ejecutar `uv add "setuptools<82.0.0"`

In [23]:
import mlflow  # importar mlflow
from sklearn.datasets import load_diabetes
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split

db = load_diabetes()
X_train, X_test, y_train, y_test = train_test_split(db.data, db.target)

# Create and train models.
rf = RandomForestRegressor(n_estimators=100, max_depth=6, max_features=3)

mlflow.autolog()  # registrar automáticamente información del entrenamiento
with mlflow.start_run():  # delimita inicio y fin del run
    # aquí comienza el run
    rf.fit(X_train, y_train)  # train the model
    predictions = rf.predict(X_test)  # Use the model to make predictions on the test dataset.
    # aquí termina el run

2026/06/10 10:54:17 INFO mlflow.tracking.fluent: Autologging successfully enabled for sklearn.
2026/06/10 10:54:17 INFO mlflow.tracking.fluent: Autologging successfully enabled for xgboost.
2026/06/10 10:54:17 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html


In [24]:
run = mlflow.last_active_run()
info = mlflow.get_run(run.info.run_id)
print(info.data.params)
print(info.data.metrics)

{'bootstrap': 'True', 'ccp_alpha': '0.0', 'criterion': 'squared_error', 'max_depth': '6', 'max_features': '3', 'max_leaf_nodes': 'None', 'max_samples': 'None', 'min_impurity_decrease': '0.0', 'min_samples_leaf': '1', 'min_samples_split': '2', 'min_weight_fraction_leaf': '0.0', 'monotonic_cst': 'None', 'n_estimators': '100', 'n_jobs': 'None', 'oob_score': 'False', 'random_state': 'None', 'verbose': '0', 'warm_start': 'False'}
{'training_mean_squared_error': 1258.685737335556, 'training_mean_absolute_error': 29.482975693714803, 'training_r2_score': 0.7896954594818133, 'training_root_mean_squared_error': 35.47796129057525, 'training_score': 0.7896954594818133}


## **1.1 Combinando Optuna + MLflow (2.0 puntos)**

Ahora que tenemos conocimiento de ambas herramientas, intentemos ahora combinarlas para **más sabor**. El objetivo de este apartado es simple: automatizar la optimización de los parámetros de nuestros modelos usando `Optuna` y registrando de forma automática cada resultado en `MLFlow`.

Considerando el objetivo planteado, se le pide completar la función `optimize_model`, la cual debe:
- **Optimizar los hiperparámetros del modelo `XGBoost` usando `Optuna`.** Realice una cantidad de iteraciones para evitar tiempos de ejecución excesivos (al menos 10)
- **Registrar cada entrenamiento en un experimento nuevo**, asegurándose de que la métrica `f1-score` se registre como `"valid_f1"`. No se deben guardar todos los experimentos en *Default*; en su lugar, cada `experiment` y `run` deben tener nombres interpretables, reconocibles y diferentes a los nombres por defecto (por ejemplo, para un run: "XGBoost con lr 0.1").
- **Devolver el mejor modelo** usando la función `get_best_model` y serializarlo en el disco con `pickle.dump`. Luego, guardar el modelo en la carpeta `/models`.
- **Guardar el código en `optimize.py`**. La ejecución de `python optimize.py` debería ejecutar la función `optimize_model`.
- **Guardar las versiones de las librerías utilizadas** en el desarrollo.

*Hint: Le puede ser útil revisar los parámetros que recibe `mlflow.start_run`*

```python
def get_best_model(experiment_id):
    runs = mlflow.search_runs(experiment_id)
    best_model_id = runs.sort_values("metrics.valid_f1")["run_id"].iloc[0]
    best_model = mlflow.sklearn.load_model("runs:/" + best_model_id + "/model")

    return best_model
```

In [ ]:
%%writefile optimize.py
import os
import pickle
import mlflow
import mlflow.sklearn
import optuna
import pandas as pd
from sklearn.metrics import f1_score
from sklearn.model_selection import train_test_split
from xgboost import XGBClassifier

df = pd.read_csv("water_potability.csv")
df = df.fillna(df.median(numeric_only=True))  # imputar nulos con mediana

X = df.drop(columns=["Potability"])
y = df["Potability"]

X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.2, random_state=42)

def get_best_model(experiment_id):
    runs = mlflow.search_runs(experiment_id)
    best_model_id = runs.sort_values("metrics.valid_f1", ascending=False)["run_id"].iloc[0]
    best_model = mlflow.sklearn.load_model("runs:/" + best_model_id + "/model")

    return best_model

def optimize_model():
    # Optimización de hiperparámetros del modelo con optuna y mlflow
    #-----------------------------------------------------------------
    # Nombre reconocible para experimento
    experiment = mlflow.get_experiment_by_name("Potabilidad_XGBoost_experimento")
    if experiment is None:
        experiment_id = mlflow.create_experiment("Potabilidad_XGBoost_experimento")
    else:
        experiment_id = experiment.experiment_id

    def objective_function(trial):
        # Definición de hiperparámetros
        params = {
            "n_estimators":  trial.suggest_int("n_estimators", 50, 400),
            "max_depth":     trial.suggest_int("max_depth", 3, 10),
            "learning_rate": trial.suggest_float("learning_rate", 0.01, 0.3),
        }

        # Nombre interpretable para el run
        run_name = f"XGBoost con lr {params['learning_rate']:.3f} y depth {params['max_depth']}"

        # Entrenamiento de XGBoost (mlflow)
        with mlflow.start_run(experiment_id=experiment_id, run_name=run_name):
            model = XGBClassifier(seed=42, eval_metric="logloss", **params)
            model.fit(
                X_train, y_train, eval_set=[(X_train, y_train), (X_valid, y_valid)],
                )
            # Prediccion y evaluacion
            yhat = model.predict(X_valid)
            valid_f1 = f1_score(y_valid, yhat)

        # Registrar resultados en mlfloww
            mlflow.log_params(params)
            mlflow.log_metric("valid_f1", valid_f1)
            mlflow.sklearn.log_model(model, name="model")

        return valid_f1

    #-----------------------------------------------------------------
    
    study = optuna.create_study(direction="maximize")
    study.optimize(objective_function, n_trials=15)

    # Obtener y guardar el mejor modelo
    best_model = get_best_model(experiment_id) # elige el con mayor f1
    os.makedirs("models", exist_ok=True) # crea carpeta models
    with open("models/best_model.pkl", "wb") as f:
        pickle.dump(best_model, f) # Guarda modelo

    print(f"Mejor F1: {study.best_value:.4f}")

    return best_model


# Guardar archivo optimize
if __name__ == "__main__":
    optimize_model()

Overwriting optimize.py


In [26]:
%run optimize.py

[I 2026-06-10 10:54:46,418] A new study created in memory with name: no-name-4f737d03-7a93-4995-a7c9-772fcc5cea5b


[0]	validation_0-logloss:0.63586	validation_1-logloss:0.64681
[1]	validation_0-logloss:0.60845	validation_1-logloss:0.63715
[2]	validation_0-logloss:0.58500	validation_1-logloss:0.63222
[3]	validation_0-logloss:0.55802	validation_1-logloss:0.62724
[4]	validation_0-logloss:0.53649	validation_1-logloss:0.62147
[5]	validation_0-logloss:0.52115	validation_1-logloss:0.62094
[6]	validation_0-logloss:0.50897	validation_1-logloss:0.61852
[7]	validation_0-logloss:0.50377	validation_1-logloss:0.61789
[8]	validation_0-logloss:0.48519	validation_1-logloss:0.61785
[9]	validation_0-logloss:0.47646	validation_1-logloss:0.61545
[10]	validation_0-logloss:0.46542	validation_1-logloss:0.61448
[11]	validation_0-logloss:0.45819	validation_1-logloss:0.61307
[12]	validation_0-logloss:0.45568	validation_1-logloss:0.61177
[13]	validation_0-logloss:0.44421	validation_1-logloss:0.60874
[14]	validation_0-logloss:0.42049	validation_1-logloss:0.60727
[15]	validation_0-logloss:0.41281	validation_1-logloss:0.60626
[1

2026/06/10 10:54:47 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:54:57 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:55:04,895] Trial 0 finished with value: 0.4597156398104265 and parameters: {'n_estimators': 138, 'max_depth': 8, 'learning_rate': 0.180381101867143}. Best is trial 0 with value: 0.4597156398104265.


[0]	validation_0-logloss:0.64180	validation_1-logloss:0.64895
[1]	validation_0-logloss:0.61940	validation_1-logloss:0.63722
[2]	validation_0-logloss:0.60071	validation_1-logloss:0.62992
[3]	validation_0-logloss:0.58755	validation_1-logloss:0.62429
[4]	validation_0-logloss:0.56395	validation_1-logloss:0.61968
[5]	validation_0-logloss:0.55144	validation_1-logloss:0.61752
[6]	validation_0-logloss:0.53646	validation_1-logloss:0.61126
[7]	validation_0-logloss:0.53113	validation_1-logloss:0.60839
[8]	validation_0-logloss:0.52243	validation_1-logloss:0.60785
[9]	validation_0-logloss:0.51607	validation_1-logloss:0.60618
[10]	validation_0-logloss:0.50433	validation_1-logloss:0.60710
[11]	validation_0-logloss:0.48487	validation_1-logloss:0.60210
[12]	validation_0-logloss:0.47932	validation_1-logloss:0.60097
[13]	validation_0-logloss:0.47204	validation_1-logloss:0.60128
[14]	validation_0-logloss:0.46987	validation_1-logloss:0.60025
[15]	validation_0-logloss:0.46777	validation_1-logloss:0.59970
[1

2026/06/10 10:55:07 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:55:16 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:55:22,832] Trial 1 finished with value: 0.4541284403669725 and parameters: {'n_estimators': 201, 'max_depth': 7, 'learning_rate': 0.1867987793874143}. Best is trial 0 with value: 0.4597156398104265.


[0]	validation_0-logloss:0.63705	validation_1-logloss:0.65205
[1]	validation_0-logloss:0.60946	validation_1-logloss:0.64309
[2]	validation_0-logloss:0.58721	validation_1-logloss:0.63876
[3]	validation_0-logloss:0.56320	validation_1-logloss:0.63276
[4]	validation_0-logloss:0.54377	validation_1-logloss:0.62799
[5]	validation_0-logloss:0.52735	validation_1-logloss:0.62553
[6]	validation_0-logloss:0.51028	validation_1-logloss:0.62284
[7]	validation_0-logloss:0.49513	validation_1-logloss:0.62053
[8]	validation_0-logloss:0.48600	validation_1-logloss:0.61784
[9]	validation_0-logloss:0.47870	validation_1-logloss:0.61721
[10]	validation_0-logloss:0.46575	validation_1-logloss:0.61675
[11]	validation_0-logloss:0.45838	validation_1-logloss:0.61527
[12]	validation_0-logloss:0.45130	validation_1-logloss:0.61426
[13]	validation_0-logloss:0.44342	validation_1-logloss:0.61285
[14]	validation_0-logloss:0.43230	validation_1-logloss:0.61094
[15]	validation_0-logloss:0.42244	validation_1-logloss:0.61063
[1

2026/06/10 10:55:25 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:55:34 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:55:39,980] Trial 2 finished with value: 0.4942263279445728 and parameters: {'n_estimators': 290, 'max_depth': 10, 'learning_rate': 0.1263628108092111}. Best is trial 2 with value: 0.4942263279445728.


[0]	validation_0-logloss:0.66138	validation_1-logloss:0.65501
[1]	validation_0-logloss:0.65218	validation_1-logloss:0.64847
[2]	validation_0-logloss:0.64460	validation_1-logloss:0.64307
[3]	validation_0-logloss:0.64123	validation_1-logloss:0.64054
[4]	validation_0-logloss:0.63532	validation_1-logloss:0.63929
[5]	validation_0-logloss:0.63127	validation_1-logloss:0.63683
[6]	validation_0-logloss:0.62849	validation_1-logloss:0.63388
[7]	validation_0-logloss:0.62511	validation_1-logloss:0.63223
[8]	validation_0-logloss:0.62224	validation_1-logloss:0.63108
[9]	validation_0-logloss:0.61921	validation_1-logloss:0.63037
[10]	validation_0-logloss:0.61776	validation_1-logloss:0.63060
[11]	validation_0-logloss:0.61238	validation_1-logloss:0.62610
[12]	validation_0-logloss:0.60666	validation_1-logloss:0.62220
[13]	validation_0-logloss:0.60173	validation_1-logloss:0.61735
[14]	validation_0-logloss:0.59940	validation_1-logloss:0.61777
[15]	validation_0-logloss:0.59863	validation_1-logloss:0.61797
[1

2026/06/10 10:55:41 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:55:50 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:55:56,027] Trial 3 finished with value: 0.49760765550239233 and parameters: {'n_estimators': 322, 'max_depth': 3, 'learning_rate': 0.20804193568591817}. Best is trial 3 with value: 0.49760765550239233.


[0]	validation_0-logloss:0.66697	validation_1-logloss:0.65854
[1]	validation_0-logloss:0.66256	validation_1-logloss:0.65525
[2]	validation_0-logloss:0.65882	validation_1-logloss:0.65276
[3]	validation_0-logloss:0.65624	validation_1-logloss:0.65119
[4]	validation_0-logloss:0.65322	validation_1-logloss:0.64953
[5]	validation_0-logloss:0.65008	validation_1-logloss:0.64726
[6]	validation_0-logloss:0.64849	validation_1-logloss:0.64660
[7]	validation_0-logloss:0.64586	validation_1-logloss:0.64552
[8]	validation_0-logloss:0.64443	validation_1-logloss:0.64435
[9]	validation_0-logloss:0.64233	validation_1-logloss:0.64282
[10]	validation_0-logloss:0.64113	validation_1-logloss:0.64191
[11]	validation_0-logloss:0.63886	validation_1-logloss:0.64080
[12]	validation_0-logloss:0.63743	validation_1-logloss:0.64002
[13]	validation_0-logloss:0.63572	validation_1-logloss:0.63916
[14]	validation_0-logloss:0.63421	validation_1-logloss:0.63809
[15]	validation_0-logloss:0.63333	validation_1-logloss:0.63767
[1

2026/06/10 10:55:57 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:56:05 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:56:11,893] Trial 4 finished with value: 0.4173441734417344 and parameters: {'n_estimators': 214, 'max_depth': 3, 'learning_rate': 0.078945998449519}. Best is trial 3 with value: 0.49760765550239233.


[0]	validation_0-logloss:0.65579	validation_1-logloss:0.65455
[1]	validation_0-logloss:0.64123	validation_1-logloss:0.64808
[2]	validation_0-logloss:0.62815	validation_1-logloss:0.64070
[3]	validation_0-logloss:0.61875	validation_1-logloss:0.63505
[4]	validation_0-logloss:0.61136	validation_1-logloss:0.63163
[5]	validation_0-logloss:0.60527	validation_1-logloss:0.62877
[6]	validation_0-logloss:0.59478	validation_1-logloss:0.62570
[7]	validation_0-logloss:0.59024	validation_1-logloss:0.62445
[8]	validation_0-logloss:0.58143	validation_1-logloss:0.62012
[9]	validation_0-logloss:0.57698	validation_1-logloss:0.61972
[10]	validation_0-logloss:0.56987	validation_1-logloss:0.61750
[11]	validation_0-logloss:0.56721	validation_1-logloss:0.61731
[12]	validation_0-logloss:0.55750	validation_1-logloss:0.61354
[13]	validation_0-logloss:0.55298	validation_1-logloss:0.61248
[14]	validation_0-logloss:0.54705	validation_1-logloss:0.61126
[15]	validation_0-logloss:0.54561	validation_1-logloss:0.61106
[1

2026/06/10 10:56:13 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:56:22 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:56:28,276] Trial 5 finished with value: 0.47417840375586856 and parameters: {'n_estimators': 278, 'max_depth': 6, 'learning_rate': 0.12450782644905452}. Best is trial 3 with value: 0.49760765550239233.


[0]	validation_0-logloss:0.62467	validation_1-logloss:0.64954
[1]	validation_0-logloss:0.58865	validation_1-logloss:0.63912
[2]	validation_0-logloss:0.56182	validation_1-logloss:0.63645
[3]	validation_0-logloss:0.53701	validation_1-logloss:0.63391
[4]	validation_0-logloss:0.50779	validation_1-logloss:0.62354
[5]	validation_0-logloss:0.48124	validation_1-logloss:0.62199
[6]	validation_0-logloss:0.46494	validation_1-logloss:0.61985
[7]	validation_0-logloss:0.44841	validation_1-logloss:0.61301
[8]	validation_0-logloss:0.43165	validation_1-logloss:0.61345
[9]	validation_0-logloss:0.41786	validation_1-logloss:0.61424
[10]	validation_0-logloss:0.40186	validation_1-logloss:0.61485
[11]	validation_0-logloss:0.39876	validation_1-logloss:0.61442
[12]	validation_0-logloss:0.39049	validation_1-logloss:0.61451
[13]	validation_0-logloss:0.37667	validation_1-logloss:0.61549
[14]	validation_0-logloss:0.37400	validation_1-logloss:0.61358
[15]	validation_0-logloss:0.36395	validation_1-logloss:0.61499
[1

2026/06/10 10:56:30 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:56:39 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:56:45,328] Trial 6 finished with value: 0.4864864864864865 and parameters: {'n_estimators': 356, 'max_depth': 10, 'learning_rate': 0.17668550620278597}. Best is trial 3 with value: 0.49760765550239233.


[0]	validation_0-logloss:0.66153	validation_1-logloss:0.65555
[1]	validation_0-logloss:0.65144	validation_1-logloss:0.64959
[2]	validation_0-logloss:0.64229	validation_1-logloss:0.64639
[3]	validation_0-logloss:0.63175	validation_1-logloss:0.63922
[4]	validation_0-logloss:0.62670	validation_1-logloss:0.63659
[5]	validation_0-logloss:0.62311	validation_1-logloss:0.63548
[6]	validation_0-logloss:0.61434	validation_1-logloss:0.63076
[7]	validation_0-logloss:0.60967	validation_1-logloss:0.62792
[8]	validation_0-logloss:0.60406	validation_1-logloss:0.62677
[9]	validation_0-logloss:0.60128	validation_1-logloss:0.62511
[10]	validation_0-logloss:0.59318	validation_1-logloss:0.62332
[11]	validation_0-logloss:0.58679	validation_1-logloss:0.62179
[12]	validation_0-logloss:0.58510	validation_1-logloss:0.62084
[13]	validation_0-logloss:0.57827	validation_1-logloss:0.61860
[14]	validation_0-logloss:0.57456	validation_1-logloss:0.61746
[15]	validation_0-logloss:0.57293	validation_1-logloss:0.61648
[1

2026/06/10 10:56:46 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:56:54 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:57:00,552] Trial 7 finished with value: 0.45454545454545453 and parameters: {'n_estimators': 159, 'max_depth': 5, 'learning_rate': 0.10381988863904716}. Best is trial 3 with value: 0.49760765550239233.


[0]	validation_0-logloss:0.65818	validation_1-logloss:0.65314
[1]	validation_0-logloss:0.64696	validation_1-logloss:0.64658
[2]	validation_0-logloss:0.63803	validation_1-logloss:0.64289
[3]	validation_0-logloss:0.63199	validation_1-logloss:0.63954
[4]	validation_0-logloss:0.62870	validation_1-logloss:0.63789
[5]	validation_0-logloss:0.61992	validation_1-logloss:0.63245
[6]	validation_0-logloss:0.61517	validation_1-logloss:0.63089
[7]	validation_0-logloss:0.61129	validation_1-logloss:0.63049
[8]	validation_0-logloss:0.60903	validation_1-logloss:0.63018
[9]	validation_0-logloss:0.60750	validation_1-logloss:0.62933
[10]	validation_0-logloss:0.60578	validation_1-logloss:0.62953
[11]	validation_0-logloss:0.60419	validation_1-logloss:0.62777
[12]	validation_0-logloss:0.60084	validation_1-logloss:0.62785
[13]	validation_0-logloss:0.59768	validation_1-logloss:0.62699
[14]	validation_0-logloss:0.58992	validation_1-logloss:0.61915
[15]	validation_0-logloss:0.58377	validation_1-logloss:0.61383
[1

2026/06/10 10:57:01 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:57:09 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:57:15,538] Trial 8 finished with value: 0.4691358024691358 and parameters: {'n_estimators': 153, 'max_depth': 3, 'learning_rate': 0.29209844805556434}. Best is trial 3 with value: 0.49760765550239233.


[0]	validation_0-logloss:0.65829	validation_1-logloss:0.65371
[1]	validation_0-logloss:0.64524	validation_1-logloss:0.64681
[2]	validation_0-logloss:0.63440	validation_1-logloss:0.64226
[3]	validation_0-logloss:0.62792	validation_1-logloss:0.63728
[4]	validation_0-logloss:0.61743	validation_1-logloss:0.63144
[5]	validation_0-logloss:0.61359	validation_1-logloss:0.63033
[6]	validation_0-logloss:0.60295	validation_1-logloss:0.62663
[7]	validation_0-logloss:0.59566	validation_1-logloss:0.62353
[8]	validation_0-logloss:0.58795	validation_1-logloss:0.61991
[9]	validation_0-logloss:0.58374	validation_1-logloss:0.61825
[10]	validation_0-logloss:0.58083	validation_1-logloss:0.61807
[11]	validation_0-logloss:0.57719	validation_1-logloss:0.61759
[12]	validation_0-logloss:0.57183	validation_1-logloss:0.61576
[13]	validation_0-logloss:0.56887	validation_1-logloss:0.61510
[14]	validation_0-logloss:0.56619	validation_1-logloss:0.61525
[15]	validation_0-logloss:0.56374	validation_1-logloss:0.61533
[1

2026/06/10 10:57:17 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:57:25 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:57:32,183] Trial 9 finished with value: 0.48226950354609927 and parameters: {'n_estimators': 293, 'max_depth': 5, 'learning_rate': 0.14297740431261732}. Best is trial 3 with value: 0.49760765550239233.


[0]	validation_0-logloss:0.66946	validation_1-logloss:0.66017
[1]	validation_0-logloss:0.66819	validation_1-logloss:0.65931
[2]	validation_0-logloss:0.66697	validation_1-logloss:0.65849
[3]	validation_0-logloss:0.66580	validation_1-logloss:0.65762
[4]	validation_0-logloss:0.66467	validation_1-logloss:0.65687
[5]	validation_0-logloss:0.66358	validation_1-logloss:0.65611
[6]	validation_0-logloss:0.66234	validation_1-logloss:0.65505
[7]	validation_0-logloss:0.66089	validation_1-logloss:0.65419
[8]	validation_0-logloss:0.65951	validation_1-logloss:0.65330
[9]	validation_0-logloss:0.65852	validation_1-logloss:0.65275
[10]	validation_0-logloss:0.65721	validation_1-logloss:0.65211
[11]	validation_0-logloss:0.65594	validation_1-logloss:0.65132
[12]	validation_0-logloss:0.65471	validation_1-logloss:0.65061
[13]	validation_0-logloss:0.65381	validation_1-logloss:0.65001
[14]	validation_0-logloss:0.65273	validation_1-logloss:0.64924
[15]	validation_0-logloss:0.65156	validation_1-logloss:0.64857
[1

2026/06/10 10:57:32 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:57:41 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:57:47,078] Trial 10 finished with value: 0.20422535211267606 and parameters: {'n_estimators': 52, 'max_depth': 4, 'learning_rate': 0.019083916967121167}. Best is trial 3 with value: 0.49760765550239233.


[0]	validation_0-logloss:0.60805	validation_1-logloss:0.64704
[1]	validation_0-logloss:0.56306	validation_1-logloss:0.63771
[2]	validation_0-logloss:0.52670	validation_1-logloss:0.63224
[3]	validation_0-logloss:0.49595	validation_1-logloss:0.62761
[4]	validation_0-logloss:0.47285	validation_1-logloss:0.62461
[5]	validation_0-logloss:0.46556	validation_1-logloss:0.62225
[6]	validation_0-logloss:0.44282	validation_1-logloss:0.62677
[7]	validation_0-logloss:0.41027	validation_1-logloss:0.62785
[8]	validation_0-logloss:0.40098	validation_1-logloss:0.62771
[9]	validation_0-logloss:0.38394	validation_1-logloss:0.62511
[10]	validation_0-logloss:0.36650	validation_1-logloss:0.62518
[11]	validation_0-logloss:0.35082	validation_1-logloss:0.62526
[12]	validation_0-logloss:0.34496	validation_1-logloss:0.62594
[13]	validation_0-logloss:0.32831	validation_1-logloss:0.62508
[14]	validation_0-logloss:0.32058	validation_1-logloss:0.62619
[15]	validation_0-logloss:0.30848	validation_1-logloss:0.62609
[1

2026/06/10 10:57:49 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:57:58 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:58:04,210] Trial 11 finished with value: 0.48758465011286684 and parameters: {'n_estimators': 400, 'max_depth': 10, 'learning_rate': 0.2485407892459203}. Best is trial 3 with value: 0.49760765550239233.


[0]	validation_0-logloss:0.62734	validation_1-logloss:0.64399
[1]	validation_0-logloss:0.59775	validation_1-logloss:0.63476
[2]	validation_0-logloss:0.57245	validation_1-logloss:0.63243
[3]	validation_0-logloss:0.55713	validation_1-logloss:0.62409
[4]	validation_0-logloss:0.52560	validation_1-logloss:0.62062
[5]	validation_0-logloss:0.51386	validation_1-logloss:0.62034
[6]	validation_0-logloss:0.49085	validation_1-logloss:0.62110
[7]	validation_0-logloss:0.47099	validation_1-logloss:0.62012
[8]	validation_0-logloss:0.46467	validation_1-logloss:0.62087
[9]	validation_0-logloss:0.45134	validation_1-logloss:0.62056
[10]	validation_0-logloss:0.44820	validation_1-logloss:0.61854
[11]	validation_0-logloss:0.43894	validation_1-logloss:0.61861
[12]	validation_0-logloss:0.43486	validation_1-logloss:0.61885
[13]	validation_0-logloss:0.42022	validation_1-logloss:0.61907
[14]	validation_0-logloss:0.40188	validation_1-logloss:0.61659
[15]	validation_0-logloss:0.39194	validation_1-logloss:0.62006
[1

2026/06/10 10:58:06 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:58:15 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:58:21,383] Trial 12 finished with value: 0.4764044943820225 and parameters: {'n_estimators': 303, 'max_depth': 8, 'learning_rate': 0.22975953788122466}. Best is trial 3 with value: 0.49760765550239233.


[0]	validation_0-logloss:0.65579	validation_1-logloss:0.65584
[1]	validation_0-logloss:0.64295	validation_1-logloss:0.65081
[2]	validation_0-logloss:0.63113	validation_1-logloss:0.64711
[3]	validation_0-logloss:0.61924	validation_1-logloss:0.64159
[4]	validation_0-logloss:0.60736	validation_1-logloss:0.63898
[5]	validation_0-logloss:0.59719	validation_1-logloss:0.63751
[6]	validation_0-logloss:0.58641	validation_1-logloss:0.63415
[7]	validation_0-logloss:0.57759	validation_1-logloss:0.63284
[8]	validation_0-logloss:0.56881	validation_1-logloss:0.62988
[9]	validation_0-logloss:0.56071	validation_1-logloss:0.62692
[10]	validation_0-logloss:0.55380	validation_1-logloss:0.62534
[11]	validation_0-logloss:0.54711	validation_1-logloss:0.62371
[12]	validation_0-logloss:0.53723	validation_1-logloss:0.62242
[13]	validation_0-logloss:0.53066	validation_1-logloss:0.62141
[14]	validation_0-logloss:0.52455	validation_1-logloss:0.62017
[15]	validation_0-logloss:0.51777	validation_1-logloss:0.61958
[1

2026/06/10 10:58:24 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:58:32 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:58:37,872] Trial 13 finished with value: 0.4728132387706856 and parameters: {'n_estimators': 342, 'max_depth': 9, 'learning_rate': 0.06223882453686948}. Best is trial 3 with value: 0.49760765550239233.


[0]	validation_0-logloss:0.64491	validation_1-logloss:0.65054
[1]	validation_0-logloss:0.62190	validation_1-logloss:0.63646
[2]	validation_0-logloss:0.60906	validation_1-logloss:0.63114
[3]	validation_0-logloss:0.58342	validation_1-logloss:0.62330
[4]	validation_0-logloss:0.56528	validation_1-logloss:0.61875
[5]	validation_0-logloss:0.55782	validation_1-logloss:0.61692
[6]	validation_0-logloss:0.55389	validation_1-logloss:0.61448
[7]	validation_0-logloss:0.54394	validation_1-logloss:0.61105
[8]	validation_0-logloss:0.53126	validation_1-logloss:0.61208
[9]	validation_0-logloss:0.52727	validation_1-logloss:0.61247
[10]	validation_0-logloss:0.51242	validation_1-logloss:0.61008
[11]	validation_0-logloss:0.50557	validation_1-logloss:0.61184
[12]	validation_0-logloss:0.49825	validation_1-logloss:0.61241
[13]	validation_0-logloss:0.49471	validation_1-logloss:0.61382
[14]	validation_0-logloss:0.49133	validation_1-logloss:0.61213
[15]	validation_0-logloss:0.48722	validation_1-logloss:0.61206
[1

2026/06/10 10:58:39 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/06/10 10:58:48 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
[I 2026-06-10 10:58:54,379] Trial 14 finished with value: 0.4764044943820225 and parameters: {'n_estimators': 271, 'max_depth': 6, 'learning_rate': 0.22583579500291528}. Best is trial 3 with value: 0.49760765550239233.

Mejor F1: 0.4976


# **2. FastAPI (2.0 puntos)**

<div align="center">
  <img src="https://media3.giphy.com/media/YQitE4YNQNahy/giphy-downsized-large.gif" width="500">
</div>

Con el modelo ya entrenado, la idea de esta sección es generar una API REST a la cual se le pueda hacer *requests* para así interactuar con su modelo. En particular, se le pide:

- Guardar el código de esta sección en el archivo `main.py`. Note que ejecutar `python main.py` debería levantar el servidor en el puerto por defecto.
- Defina `GET` con ruta tipo *home* que describa brevemente su modelo, el problema que intenta resolver, su entrada y salida.
- Defina un `POST` a la ruta `/potabilidad/` donde utilice su mejor optimizado para predecir si una medición de agua es o no potable. Por ejemplo, una llamada de esta ruta con un *body*:

```json
{
   "ph":10.316400384553162,
   "Hardness":217.2668424334475,
   "Solids":10676.508475429378,
   "Chloramines":3.445514571005745,
   "Sulfate":397.7549459751925,
   "Conductivity":492.20647361771086,
   "Organic_carbon":12.812732207582542,
   "Trihalomethanes":72.28192021570328,
   "Turbidity":3.4073494284238364
}
```

Su servidor debería retornar una respuesta HTML con código 200 con:


```json
{
  "potabilidad": 0 # respuesta puede variar según el clasificador que entrenen
}
```

**`HINT:` Recuerde que puede utilizar [http://localhost:8000/docs](http://localhost:8000/docs) para hacer un `POST`.**

In [ ]:
%%writefile main.py
import pickle
import uvicorn
import pandas as pd
from fastapi import FastAPI
from pydantic import BaseModel

# Cargar modelo
with open("models/best_model.pkl", "rb") as f:
    model = pickle.load(f)

# Definir estructura de entrada
class WaterMeasurement(BaseModel):
    ph: float
    Hardness: float
    Solids: float
    Chloramines: float
    Sulfate: float
    Conductivity: float
    Organic_carbon: float
    Trihalomethanes: float
    Turbidity: float

# Crear app
app = FastAPI()

# GET home
@app.get("/")
def home():
    return {
        "modelo": "XGBoost optimizado con Optuna",
        "problema": "Clasificación binaria de potabilidad del agua",
        "entrada": "9 mediciones químicas: ph, Hardness, Solids , Chloramines, Sulfate, Conductivity, Organic_carbon, Trihalomethanes, Turbidity",
        "salida": "potabilidad: 1 (potable) o 0 (no potable)"
    }

# POST predicción
@app.post("/potabilidad/")
def predecir_potabilidad(medicion: WaterMeasurement):
    datos = pd.DataFrame([medicion.model_dump()])
    prediccion = model.predict(datos)[0]
    return {"potabilidad": int(prediccion)}

if __name__ == "__main__":
    import nest_asyncio
    nest_asyncio.apply()
    uvicorn.run(app, host="0.0.0.0", port=8000)

Overwriting main.py


In [33]:
import subprocess
import sys

subprocess.Popen([sys.executable, "-m", "uvicorn", "main:app", "--host", "0.0.0.0", "--port", "8000"])

<Popen: returncode: None args: ['/Users/javierayanez/Documents/Semestre 11/M...>

INFO:     Started server process [16780]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)


INFO:     127.0.0.1:52530 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:52530 - "GET /favicon.ico HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:52533 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:52533 - "GET /favicon.ico HTTP/1.1" 404 Not Found
INFO:     127.0.0.1:52538 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:52538 - "GET /openapi.json HTTP/1.1" 200 OK
INFO:     127.0.0.1:52540 - "POST /potabilidad/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:52540 - "POST /potabilidad/ HTTP/1.1" 200 OK
INFO:     127.0.0.1:52552 - "GET /docs HTTP/1.1" 200 OK
INFO:     127.0.0.1:52552 - "GET /openapi.json HTTP/1.1" 200 OK


# **3. Docker (2 puntos)**

<div align="center">
  <img src="https://miro.medium.com/v2/resize:fit:1400/1*9rafh2W0rbRJIKJzqYc8yA.gif" width="500">
</div>

Tras el éxito de su aplicación web para generar la salida, Smapina le solicita que genere un contenedor para poder ejecutarla en cualquier computador de la empresa de agua potable.

## **3.1 Creación de Container (1 punto)**

Cree un Dockerfile que use una imagen base de Python, copie los archivos del proyecto e instale las dependencias desde un `requirements.txt`. Con esto, construya y ejecute el contenedor Docker para la API configurada anteriormente. Entregue el código fuente (incluyendo `main.py`, `requirements.txt`, y `Dockerfile`) y la imagen Docker de la aplicación. Para la dockerización, asegúrese de cumplir con los siguientes puntos:

1. **Generar un archivo `.dockerignore`** que ignore carpetas y archivos innecesarios dentro del contenedor.
2. **Configurar un volumen** que permita la persistencia de los datos en una ruta local del computador.
3. **Exponer el puerto** para acceder a la ruta de la API sin tener que entrar al contenedor directamente.
4. **Incluir imágenes en el notebook** que muestren la ejecución del contenedor y los resultados obtenidos.
5. **Revisar y comentar los recursos utilizados por el contenedor**. Analice si los contenedores son livianos en términos de recursos.

## **3.2 Preguntas de Smapina (1 punto)**
Tras haber experimentado con Docker, Smapina desea profundizar más en el tema y decide realizarle las siguientes consultas:

- ¿Cómo se diferencia Docker de una máquina virtual (VM)?
- ¿Cuál es la diferencia entre usar Docker y ejecutar la aplicación directamente en el sistema local?
- ¿Cómo asegura Docker la consistencia entre diferentes entornos de desarrollo y producción?
- ¿Cómo se gestionan los volúmenes en Docker para la persistencia de datos?
- ¿Qué son Dockerfile y docker-compose.yml, y cuál es su propósito?

In [38]:
%%writefile Dockerfile
FROM python:3.11-slim

WORKDIR /app

COPY requirements.txt .
RUN pip install --no-cache-dir -r requirements.txt

COPY main.py .
COPY models/ models/

EXPOSE 8000

CMD ["python", "main.py"]

Overwriting Dockerfile


In [41]:
%%writefile .dockerignore
.venv
__pycache__
*.pyc
.git
mlruns
mlflow.db
*.ipynb
.ipynb_checkpoints
optimize.py
water_potability.csv

Overwriting .dockerignore


In [50]:
!docker ps

CONTAINER ID   IMAGE                  COMMAND            CREATED         STATUS         PORTS                                         NAMES
bceebd90e3af   water-potability-api   "python main.py"   2 minutes ago   Up 2 minutes   0.0.0.0:8000->8000/tcp, [::]:8000->8000/tcp   water-api


In [53]:
!docker build -t water-potability-api .
!docker run -d -p 8000:8000 -v $(pwd)/models:/app/models --name water-api water-potability-api
!docker stats water-api --no-stream





[+] Building 0.0s (0/1)                                    docker:desktop-linux
[+] Building 0.2s (1/2)                                    docker:desktop-linux
 => [internal] load build definition from Dockerfile                       0.0s
 => => transferring dockerfile: 228B                                       0.0s
 => [internal] load metadata for docker.io/library/python:3.11-slim        0.2s
[+] Building 0.3s (1/2)                                    docker:desktop-linux
 => [internal] load build definition from Dockerfile                       0.0s
 => => transferring dockerfile: 228B                                       0.0s
 => [internal] load metadata for docker.io/library/python:3.11-slim        0.3s
[+] Building 0.5s (1/2)                                    docker:desktop-linux
 => [internal] load build definition from Dockerfile                       0.0s
 => => transferring dockerfile: 228B                                       0.0s
 => [internal] load metadata for doc

![docker ps](ss2.png)

![docker stats](ss1.png)

![fastapi docs](ss3.png)

![post result](ss4.png)

Respuesta 3.1: 
> Según las capturas con los resultados de ejecución, el contenedor es bastante liviano en comparación con una VM.  Usa solo 138 MB de RAM para correr toda la API con el modelo XGBoost,  mientras que una VM requeriría varios GB solo para el sistema operativo.  El uso de CPU es prácticamente nulo (0.43%) en reposo, lo que confirma que  Docker es una solucion eficiente para desplegar aplicaciones como esta. Además, se configuró un volumen -v $(pwd)/models:/app/models para que el modelo persista localmente.

Respuesta 3.2:
> 1. Una maquina virtual VM virtualiza el hardware entero, es decir, instala un sistema operativo entero dentro de otro lo que puede ser muy pesado y lento. En cambio, Docker aisla la app a nivel proceso, no duplica el sistema operativo, sino que comparte el kernel de la maquina real, lo que lo hace mas liviano y rapido.
> 2. Hacerlo de manera local hace que dependa del entrono del computador, por lo que si algo se actualiza o cambia en la configuración, la app puede romperse. En cambio, Docker "empaqueta" la app con todas sus dependencias incluidas, lo que hace que pueda funcionar de forma aislada, permitiendo ser utilizado desde cualquier computador.
> 3. Docker mantiene la consistencia mediante el uso de imagenes (que no cambian), y como no cambia, a la hora de ejecución, será igual que cuando el programador lo ejecutó en su propia computadora (se utiliza una copia identica).
> 4. Los volumenes en Docker permiten que los datos persistan aunque el contenedor se detenga o elimine,  montando una carpeta del sistema local dentro del contenedor.
> 5. Dockerfile es un archivo de texto con las instrucciones para la construccion de la imagen , definiendo librerias, sistema base, etc.; y docker compose.yml es una archivo de configuracion para orquestar multiples contenedores a la vez, definiendo como se comunican y permitiendo encenderlo (ejemplo, comunicar la app del pc , una base de datos y mlflow).

# Conclusión

Éxito!
<div align="center">
  <img src="https://i.pinimg.com/originals/55/f5/fd/55f5fdc9455989f8caf7fca7f93bd96a.gif" width="500">
</div>